### PR TITLE
Add basic pieces for an initial charm

### PR DESCRIPTION
**Pull request by plars**
This is currently pushed to charmhub and can be installed from there. It works, in so much as you can relate a mongodb charm to it and it can connect to the database. However, I haven't added ingress support yet. That's still coming.

### CODE REVIEW

File: charm/metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+containers:
+  testflinger:
+    resource: testflinger-image
+
+resources:
+  testflinger-image:
+    type: oci-image
+    description: OCI image for the 'testflinger' container
+    upstream-source: plars/testflinger:latest
```
Comment (nadzyah) [inline charm/metadata.yaml:16]: Is there any canonical testflinger image? Or are we okay with using yours?

File: charm/src/charm.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        self.framework.observe(
+            self.on.testflinger_pebble_ready, self._on_testflinger_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+    @property
+    def version(self) -> str:
+        """Report the current version of the app"""
+        # TODO: get the version somehow and return it
+        return "Version ?"
```
Comment (nadzyah) [inline charm/src/charm.py:56]: What if we get it from a docker image version? Is it possible?

File: charm/src/charm.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        initial = dict(self._stored.reldata["mongodb"])
+        self._stored.reldata["mongodb"].update(
+            self.mongodb.fetch_relation_data()[event.relation.id]
+        )
+        if initial != self._stored.reldata["mongodb"]:
+            self._update_layer_and_restart(None)
+
+    def _update_layer_and_restart(self, event) -> None:
+        """Define and start layer for testflinger using Pebble"""
+
```
Comment (nadzyah) [inline charm/src/charm.py:83]: I guess this blank line is redundant. Have you run black on these files?

File: charm/src/charm.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        # add workload version in juju status
+        self.unit.set_workload_version(self.version)
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_mongodb_client_relation_removed(
+        self, event: ops.framework.EventBase
+    ) -> None:
+        """Event is fired when relation with mongodb is broken."""
+        self.unit.status = ops.WaitingStatus("Waiting for database relation")
+        raise SystemExit(0)
```
Comment (nadzyah) [inline charm/src/charm.py:114]: `sys.exit(0)` is a preferable way to raise this exception. See https://docs.python.org/3/library/exceptions.html#SystemExit

File: charm/src/charm.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    @property
+    def app_environment(self) -> dict:
+        """Get dict of env data for the mongodb credentials"""
+        db_data = self.fetch_mongodb_relation_data()
+        env = {
+            "MONGODB_HOST": db_data.get("db_host", None),
+            "MONGODB_PORT": db_data.get("db_port", None),
+            "MONGODB_USERNAME": db_data.get("db_username", None),
+            "MONGODB_PASSWORD": db_data.get("db_password", None),
+            "MONGODB_DATABASE": db_data.get("db_database", None),
```
Comment (nadzyah) [inline charm/src/charm.py:155]: It’s None by default, so you don't need to specify it explicitly

**Review by nadzyah — CHANGES_REQUESTED** [review]
There are just minor things I’ve noticed in the code. However, in general, it looks excellent (as usual☺️)

File: charm/metadata.yaml (hunk 1 above)
Comment (plars) [reply charm/metadata.yaml:16]: Not at this point, but it probably won't matter as there doesn't seem to be a way that I've found to force it to pull unless you specify a new tag every time. It always wants to use a cached copy. However, this doesn't seem to be the way to get the image anyway. When you upload the charm to the charm store, you also have to manually upload any resources (including the image) or else it won't work. So it won't actually get it from here anyway.

Also, from the juju documentation it tells us that this field is ignored by juju. It's just for reference.

File: charm/src/charm.py (hunk 2 above)
Comment (plars) [reply charm/src/charm.py:56]: Not from what I can tell. We should probably add some sort of get version api in testflinger itself. It's not critical but I liked this idea from the examples I saw and wanted to leave it there as a placeholder in the charm.

File: charm/src/charm.py (hunk 3 above)
Comment (plars) [reply charm/src/charm.py]: I do run black but I don't think it complained about this. I removed the line though.

File: charm/src/charm.py (hunk 4 above)
Comment (plars) [reply charm/src/charm.py]: done

File: charm/src/charm.py (hunk 5 above)
Comment (plars) [reply charm/src/charm.py]: done

**Review by nadzyah — APPROVED** [review]
Thanks for answering to all my questions! +1

**Pull request merged**